sources/org.osbuild.skopeo

```diff
@@ -103,17 +111,43 @@ class SkopeoSource(sources.SourceService):
                            check=True)
 
             # Verify that the digest supplied downloaded the correct container image id.
-            # The image id is the digest of the config, but skopeo can' currently
+            # The image id is the digest of the config, but skopeo can't currently
             # get the config id, only the full config, so we checksum it ourselves.
             res = subprocess.check_output(["skopeo", "inspect", "--raw", "--config", destination])
             downloaded_id = "sha256:" + hashlib.sha256(res).hexdigest()
             if downloaded_id != image_id:
                 raise RuntimeError(
                     f"Downloaded image {imagename}@{digest} has a id of {downloaded_id}, but expected {image_id}")
 
-            # Atomically move download dir into place on successful download
+            if archive_format == "dir":
+                # fetch the manifest and merge it into the archive
+                self.merge_manifest(source, dest_path)
+
+            # Atomically move download archive into place on successful download
             with ctx.suppress_oserror(errno.ENOTEMPTY, errno.EEXIST):
-                os.rename(archive_dir, f"{self.cache}/{image_id}")
+                os.makedirs(f"{self.cache}/{image_id}", exist_ok=True)
+                os.rename(f"{archive_dir}/{archive_name}", f"{self.cache}/{image_id}/{archive_name}")
+
+    def exists(self, checksum, desc):
+        archive_name = containers.archive_name(desc["image"].get("format", "docker-archive"))
+        return os.path.exists(f"{self.cache}/{checksum}/{archive_name}")
+
+    def merge_manifest(self, source, destination):
+        with tempfile.TemporaryDirectory(prefix="tmp-download-", dir=self.cache) as indexdir:
+            # download the manifest(s) to a temporary directory
+            subprocess.run(["skopeo", "copy", "--multi-arch=index-only", source, f"dir:{indexdir}"],
+                           encoding="utf-8", check=True)
+
+            # calculate the checksum of the manifest of the container image in the destination
+            manifest_path = os.path.join(destination, "manifest.json")
+            with open(manifest_path, encoding="utf-8") as manifest:
+                manifest_checksum = hashlib.sha256(manifest.read().encode()).hexdigest()
```

Comment:
Mostly pedantically, this should use `skopeo manifest-digest` instead of raw SHA256 to compute the right digest for schema1 images.

OTOH schema1 images can’t be multi-arch, so this would only result in a different manifest path for a file that would be ignored (just like this code creates `*.manifest.json` files that are ignored for other non-multiarch images).

Reply:
Still, best to use the tooling when it exists.